### ISSUE TITLE
pgBouncer not coming up after a pgo upgrade

### ISSUE THREAD

**Issue by v3nturetheworld** [CrunchyData/postgres-operator#1682]
**Describe the bug**

After upgrading the postures-operator, I tried upgrading a postgres cluster with `pgo upgrade my-postgres-cluster --namespace testproject`. After the upgrade cycle completes everything is up and running correctly aside from the pgBouncer. During and after the upgrade the pgBouncer service remains, but there is no deployment for the bouncer. Attempting to scale up or down the number of bouncers doesn't have any results, the only way I managed to get the bouncer added after the upgrade was by doing a `pgo delete pgbouncer` and then a `pgo create pgbouncer`. Note, the upgrade here is just run with 4.3.2 to 4.3.2, there aren't any actual version changes. 

**To Reproduce**
```
pgo create cluster crunchy-postgresql --cpu 15m --memory 690Mi --memory-limit 2Gi --pgbackrest-cpu 30m --pgbackrest-memory 20Mi --pgbackrest-memory-limit 60Mi --pgbouncer --pgbouncer-replicas 2 --pgbouncer-memory 20Mi --pgbouncer-memory-limit 60Mi --password <REDACTED> -n testproject 
```

Wait for deployment to complete

```
pgo show crunchy-postgresql -n testproject

sh-4.2$ pgo show cluster crunchy-postgresql -n testproject

cluster : crunchy-postgresql (crunchy-postgres-ha:centos7-12.3-4.3.2)
        pod : crunchy-postgresql-77cd86d67b-4nb98 (Running) on node2.mycluster.openshift (1/1) (primary)
        pvc : crunchy-postgresql
        pod : crunchy-postgresql-pgbouncer-7fb6d6fd68-pgqx8 (Running) on mycluster.openshift (1/1) (pgbouncer)
        pod : crunchy-postgresql-wzpx-74d7994d6b-k75kz (Running) on node2.mycluster.openshift (1/1) (replica)
        pvc : crunchy-postgresql-wzpx
        resources : CPU: 15m Memory: 690Mi
        limits : Memory: 2Gi
        storage : Primary=1Gi Replica=1Gi
        deployment : crunchy-postgresql
        deployment : crunchy-postgresql-backrest-shared-repo
        deployment : crunchy-postgresql-pgbouncer
        deployment : crunchy-postgresql-wzpx
        service : crunchy-postgresql - ClusterIP (172.30.166.250)
        service : crunchy-postgresql-pgbouncer - ClusterIP (172.30.211.79)
        service : crunchy-postgresql-replica - ClusterIP (172.30.123.59)
        pgreplica : crunchy-postgresql-wzpx
        labels : crunchy-pgha-scope=crunchy-postgresql crunchy_collect=false custom-config=pgo-custom-pg-config deployment-name=crunchy-postgresql pg-cluster=crunchy-postgresql workflowid=ccf7e7de-55f0-4597-b69c-2a76fbb11408 autofail=true crunchy-pgbadger=false name=crunchy-postgresql pg-pod-anti-affinity= pgo-backrest=true pgo-version=4.3.2 pgouser=admin
```

**Now I upgrade the cluster**

```
pgo upgrade crunchy-postgresql -n testproject
```

After waiting for the backrest backup to complete:

```
sh-4.2$ pgo show crunchy-postgresql -n testproject
cluster : crunchy-postgresql (crunchy-postgres-ha:centos7-12.3-4.3.2)
        pod : crunchy-postgresql-78cd45bbb-ntzcv (Running) on node2.mycluster.openshift (1/1) (primary)
        pvc : crunchy-postgresql
        pod : crunchy-postgresql-kyws-8db67f687-qs7gc (Running) on node1.mycluster.openshift (0/1) (unknown)
        pvc : crunchy-postgresql-kyws
        resources : CPU: 15m Memory: 690Mi
        limits : Memory: 2Gi
        storage : Primary=1Gi Replica=1Gi
        deployment : crunchy-postgresql
        deployment : crunchy-postgresql-backrest-shared-repo
        deployment : crunchy-postgresql-kyws
        service : crunchy-postgresql - ClusterIP (172.30.40.186)
        service : crunchy-postgresql-pgbouncer - ClusterIP (172.30.99.126)
        service : crunchy-postgresql-replica - ClusterIP (172.30.179.166)
        pgreplica : crunchy-postgresql-kyws
        labels : workflowid=48301e98-ce39-486a-93a7-ec2f6ff4c0a0 custom-config=pgo-custom-pg-config pg-cluster=crunchy-postgresql pgo-backrest=true pgouser=admin deployment-name=crunchy-postgresql name=crunchy-postgresql pg-pod-anti-affinity= pgo-version=4.3.2 autofail=true crunchy-pgbadger=false crunchy-pgha-scope=crunchy-postgresql crunchy_collect=false
```

**postgres operator logs**
```
time="2020-07-01T19:52:48Z" level=debug msg="set pgbouncer password in PostgreSQL" func="operator/cluster.setPostgreSQLPassword()" file="operator/cluster/pgbouncer.go:850" version=4.3.2
time="2020-07-01T19:52:48Z" level=debug msg="set PostgreSQL password for user [pgbouncer]" func="util.SetPostgreSQLPassword()" file="util/cluster.go:265" version=4.3.2
time="2020-07-01T19:52:48Z" level=debug msg="Request URL: https://172.30.0.1:443/api/v1/namespaces/testproject/pods/crunchy-postgresql-78cd45bbb-ntzcv/exec?command=psql&command=-p&command=5432&container=database&stderr=true&stdin=true&stdout=true " func="kubeapi.ExecToPodThroughAPI()" file="kubeapi/exec.go:60" version=4.3.2
time="2020-07-01T19:52:49Z" level=error msg="ERROR:  cannot execute ALTER ROLE in a read-only transaction\n" func="util.SetPostgreSQLPassword()" file="util/cluster.go:290" version=4.3.2
time="2020-07-01T19:52:49Z" level=error msg="ERROR:  cannot execute ALTER ROLE in a read-only transaction\n" func="operator/cluster.setPostgreSQLPassword()" file="operator/cluster/pgbouncer.go:859" version=4.3.2
time="2020-07-01T19:52:49Z" level=warning msg="ERROR:  cannot execute ALTER ROLE in a read-only transaction\n" func="controller/pod.(*Controller).handleBootstrapInit()" file="controller/pod/inithandler.go:179" version=4.3.2
```

**postgres pg_log**

```
< 2020-07-01 19:52:49.223 UTC [unknown] [unknown] [unknown] 00000 [local] 374 2020-07-01 19:52:49.223 UTC 5efce991.176 2020-07-01 19:52:49 UTC: >LOG:  connection received: host=[local]
< 2020-07-01 19:52:49.224 UTC [unknown] postgres postgres 00000 [local] 374 2020-07-01 19:52:49.224 UTC 5efce991.176 2020-07-01 19:52:49 UTC: >LOG:  connection authorized: user=postgres database=postgres application_name=psql
< 2020-07-01 19:52:49.232 UTC psql postgres postgres 25006 [local] 374 2020-07-01 19:52:49.232 UTC 5efce991.176 2020-07-01 19:52:49 UTC: >ERROR:  cannot execute ALTER ROLE in a read-only transaction
< 2020-07-01 19:52:49.232 UTC psql postgres postgres 25006 [local] 374 2020-07-01 19:52:49.232 UTC 5efce991.176 2020-07-01 19:52:49 UTC: >STATEMENT:  ALTER ROLE "pgbouncer" PASSWORD '<REDACTED>' LOGIN;
< 2020-07-01 19:52:49.232 UTC psql postgres postgres 00000 [local] 374 2020-07-01 19:52:49.232 UTC 5efce991.176 2020-07-01 19:52:49 UTC: >LOG:  disconnection: session time: 0:00:00.009 user=postgres database=postgres host=[localal]
```

**Expected behavior**
After an upgrade of a crunchy postgres cluster, the pgbouncer should be deployed and running 

**Screenshots**

**Please tell us about your environment:**

* Operating System: RHEL7
* Where is this running ( Local, Cloud Provider): 3 Node Cluster
* Storage being used (NFS, Hostpath, Gluster, etc): OpenEBS-SC for backrest, local storage for database
* Container Image Tag:  centos7-4.3.2
* PostgreSQL Version: 12.3
* Platform (Docker, Kubernetes, OpenShift): OpenShift
* Platform Version: 3.11

**Additional context**

**Comment by jkatz** [comment]
Hm, it looks like the Operator is trying to set up the pgbouncer user's credentials, but it is doing so on an instance that is _not_ the primary. So likely have to figure out why it thinks that instance is the primary. Seems like it's happenstance, but if you inspect the labels on the PostgreSQL instances, does there appear to be "multiple primaries?" I'll take a look at the code.

(As a note for the future, event though you have the hashed MD5 password in the logs, this is effectively the same thing as [sharing the password](https://www.slideshare.net/jkatz05/get-your-insecure-postgresql-passwords-to-scram#56), so I would recommend redacting it. I went ahead and did so above).

**Comment by jkatz** [comment]
@andrewlecuyer and I explored a few options and were unable to directly reproduce this issue. However, I was able to trigger a panic in the Operator while running similar commands, which if the Operator does panic, all execution will stop at that point. This could be the cause of the above issue, but we have yet to definitely rule on it.

The panic is fixed in #1707 and will be backpatched down to 4.3. Once committed I will cause this closed for now, but if it reoccurs after the fix is applied, we can take another look.

Thanks!

**Comment by v3nturetheworld** [comment]
Awesome, thanks for looking into this. Thanks for pointing out the md5 pw hash in the logs/screenshots I included, though I have been frequently redeploying test clusters so I was as concerned with what was included as I should probably be. I'll let you know if I see any issues with the bouncer/operator panics again

**Comment by jkatz** [comment]
Great, thanks @v3nturetheworld.

For more fun PostgreSQL facts, we did add support for SCRAM for the regular PostgreSQL user accounts in the Operator, in which if you shared me said hash, it would be much more difficult for me to guess the password. We did not add that support for pgBouncer just yet, but with pgBouncer 1.14 adding a pretty solid SCRAM implementation, we can add that in an upcoming release.
